### PR TITLE
Generate service runner implementations

### PR DESCRIPTION
**Pull request by gs-ssh16**
Changes:

- Make service execution classes implement [ServiceRunner](https://github.com/finos/legend-engine/blob/master/legend-engine-language-pure-dsl-service-execution/src/main/java/org/finos/legend/engine/language/pure/dsl/service/execution/ServiceRunner.java) interface
- Add execution class references to ServiceRunner provider-configuration file (META-INF/services/...) so that they can be loaded using ServiceLoader

### CODE REVIEW

File: legend-sdlc-generation-service/src/test/java/org/finos/legend/sdlc/generation/service/TestServiceRunnerExecution.java
Comment (kevin-m-knight-gs) [inline legend-sdlc-generation-service/src/test/java/org/finos/legend/sdlc/generation/service/TestServiceRunnerExecution.java:63]: Why is there a separate class for this? Why isn't this added to TestServiceExecutionGenerator?

File: legend-sdlc-generation-service/src/main/java/org/finos/legend/sdlc/generation/service/ServiceExecutionClassGenerator.java
Comment (kevin-m-knight-gs) [inline legend-sdlc-generation-service/src/main/java/org/finos/legend/sdlc/generation/service/ServiceExecutionClassGenerator.java:328]: I don't think there should be any calls here that create new threads. Honestly, I think AbstractServicePlanExecutor should not have any methods that create new threads. Questions of threading should be left up to the caller, not the service execution classes.

**Review by kevin-m-knight-gs — CHANGES_REQUESTED** [review]

File: legend-sdlc-generation-service/src/main/java/org/finos/legend/sdlc/generation/service/ServiceExecutionClassGenerator.java
Comment (gs-ssh16) [reply legend-sdlc-generation-service/src/main/java/org/finos/legend/sdlc/generation/service/ServiceExecutionClassGenerator.java]: Done

File: legend-sdlc-generation-service/src/main/java/org/finos/legend/sdlc/generation/service/ServiceExecutionClassGenerator.java
Comment (gs-ssh16) [reply legend-sdlc-generation-service/src/main/java/org/finos/legend/sdlc/generation/service/ServiceExecutionClassGenerator.java]: Done

File: legend-sdlc-generation-service/src/main/java/org/finos/legend/sdlc/generation/service/ServiceExecutionClassGenerator.java
Comment (gs-ssh16) [reply legend-sdlc-generation-service/src/main/java/org/finos/legend/sdlc/generation/service/ServiceExecutionClassGenerator.java]: Done

File: legend-sdlc-generation-service/src/test/java/org/finos/legend/sdlc/generation/service/TestServiceRunnerExecution.java
Comment (gs-ssh16) [reply legend-sdlc-generation-service/src/test/java/org/finos/legend/sdlc/generation/service/TestServiceRunnerExecution.java:63]: TestServiceExecutionGenerator tests 'execute' methods which are specific to each service. I added this class to test execution of different services using the common ServiceRunner API and also do some asserts on code generated exception messages (in run method). One other difference is that I wanted to improve readability of the test by using grammar to define the model instead of JSON.

File: legend-sdlc-generation-service/src/main/java/org/finos/legend/sdlc/generation/service/ServiceExecutionClassGenerator.java
Comment (gs-ssh16) [reply legend-sdlc-generation-service/src/main/java/org/finos/legend/sdlc/generation/service/ServiceExecutionClassGenerator.java]: The plan is to have the 'run' method with output stream execute in an async fashion by default and hence the thread creation. I'll discuss regarding this with @pierredebelen to see if it makes sense and get it changed in AbstractServicePlanExecutor.

File: legend-sdlc-generation-service/src/main/java/org/finos/legend/sdlc/generation/service/ServiceExecutionClassGenerator.java
Comment (gs-ssh16) [reply legend-sdlc-generation-service/src/main/java/org/finos/legend/sdlc/generation/service/ServiceExecutionClassGenerator.java:94]: This was done to ensure that this is not a breaking change (unless this change goes in, generated execution classes would fail to compile due to missing run method implementation). I plan to to have AbstractServicePlanExecutor implement ServiceRunner going forward.

File: legend-sdlc-generation-service/src/main/java/org/finos/legend/sdlc/generation/service/ServiceExecutionClassGenerator.java
Comment (kevin-m-knight-gs) [reply legend-sdlc-generation-service/src/main/java/org/finos/legend/sdlc/generation/service/ServiceExecutionClassGenerator.java]: It's much better to let the user call it asynchronously if they want that. That way the calling application manages the threads, which is more appropriate. And it's not like it's difficult for callers to call it asynchronously. Indeed, it is quite easy.

On top of that, the current approach gives no feedback on when the execution finishes or whether it finishes successfully. And if an error is thrown during the execution, it is lost to the caller. If we leave it to the caller whether to execute asynchronously, they can use an ExecutorService which will solve all those problems, as well as allowing them to manage the threads.

File: legend-sdlc-generation-service/src/test/java/org/finos/legend/sdlc/generation/service/TestServiceRunnerExecution.java
Comment (kevin-m-knight-gs) [reply legend-sdlc-generation-service/src/test/java/org/finos/legend/sdlc/generation/service/TestServiceRunnerExecution.java:63]: TestServiceExecutionGenerator is intended to test that ServiceExecutionGenerator behaves correctly. That includes the generated 'execute' methods which are specific to each service, but is not limited to that. If there are new API methods being added, they should also be tested in TestServiceExecutionGenerator. And if you want to improve readability and add more service definitions to be tested, by all means do so. But it's better to improve the existing test, rather than adding a new test.

File: legend-sdlc-generation-service/src/main/java/org/finos/legend/sdlc/generation/service/ServiceExecutionClassGenerator.java
Comment (kevin-m-knight-gs) [reply legend-sdlc-generation-service/src/main/java/org/finos/legend/sdlc/generation/service/ServiceExecutionClassGenerator.java:94]: It will be better to get AbstractServicePlanExecutor into the proper form before we release this to users. Is it as simple as adding a default implementation of run to AbstractServicePlanExecutor?

File: legend-sdlc-generation-service/src/test/java/org/finos/legend/sdlc/generation/service/TestServiceRunnerExecution.java
Comment (gs-ssh16) [reply legend-sdlc-generation-service/src/test/java/org/finos/legend/sdlc/generation/service/TestServiceRunnerExecution.java]: Done

File: legend-sdlc-generation-service/src/main/java/org/finos/legend/sdlc/generation/service/ServiceExecutionClassGenerator.java
Comment (gs-ssh16) [reply legend-sdlc-generation-service/src/main/java/org/finos/legend/sdlc/generation/service/ServiceExecutionClassGenerator.java]: Should be addressed in https://github.com/finos/legend-engine/pull/356

File: legend-sdlc-generation-service/src/main/java/org/finos/legend/sdlc/generation/service/ServiceExecutionClassGenerator.java
Comment (gs-ssh16) [reply legend-sdlc-generation-service/src/main/java/org/finos/legend/sdlc/generation/service/ServiceExecutionClassGenerator.java]: Changing it here https://github.com/finos/legend-engine/pull/356

**Comment by gs-ssh16** [comment]
Build will fail till https://github.com/finos/legend-engine/pull/356 is released

**Comment by kevin-m-knight-gs** [comment]
Once the legend-engine version is updated to include a release with https://github.com/finos/legend-engine/pull/356, this can be approved and merged (assuming, of course, all tests pass).

**Review by kevin-m-knight-gs — APPROVED** [review]

**Pull request merged**
